docs/proposals/17-ratelimits-for-shoot-dns-entries.md

```diff
@@ -0,0 +1,112 @@
+    - [Gardener](#gardener)
+  - [Alternatives](#alternatives)
+
+## Summary
+
+This `GEP` introduces rate limits for `DNSEntries` created with the help of the [shoot-dns-service](https://github.com/gardener/gardener-extension-shoot-dns-service) to protect the [dns-controller-manager](https://github.com/gardener/external-dns-management) from being throttled, i.e to protect the DNS management of Gardener from intentional or unintentional *DoS* attacks.
+
+## Motivation
+
+The `dns-controller-manager` runs update requests for DNS record on the configured DNS cloud service(s) like `AWS Route 53`. Typically these DNS cloud services are rate limiting requests on account level. As the `shoot-dns-service` allows to request and update `DNSEntries` directly or indirectly from resources on the shoot cluster, a single shoot cluster can potentially eat up the available request rate of the backend account. The `dns-controller-manager` runs as extension on each seed of a Gardener landscape and they all have to share the account for the DNS cloud service in form of a `DNSProvider` as all shoots use the same default *hosted DNS zone* for `DNSEntries` created by Gardener itself ( `external`, `ingress`) and all others created by the `shoot-dns-service`. It is this default `DNSProvider` which needs to be protected at first place. If the user provides an own `DNSProvider` and uses his hosted DNS zone(s) of his own account, the throttling would only affect this account. But also in this case, the user may want to avoid that one of his shoot clusters can break all his others clusters using the same `DNSProvider`.
+
+### Goals
+
+- The rate limit is set and applied separatedly for each `DNSProvider`
+- All update requests caused by `DNSEntries` created or updated from the `shoot-dns-service` should only be forwarded to the backend DNS service, if the update requests are within rate limit of the `DNSProvider` the `DNSEntry` is assigned to.
+- Deletion of `DNSEntries` is not rate limited to avoid freezing on shoot cluster deletion.
+- Updates of `DNSEntries` created by Gardener (e.g. `internal`, `external`, and `ingress`) should not be blocked by the rate limits.
+- If creation/updates of `DNSEntries` of the `shoot-dns-service` are pending because of a rate limit, this should also made be visible in the shoot cluster (either by events or status update)
+
+### Non-Goals
+
+- prioritize updates between `DNSEntries` assigned to the same `DNSProvider`
+
+## Proposal
+
+This proposal involves changes in two components:
+
+- [dns-controller-manager](https://github.com/gardener/external-dns-management) for implementation of rate limiting
+
+- [gardener](https://github.com/gardener/gardener) for configuration of rate limits, applying them to the external `DNSProvider` and annotating `DNSEntries` to exclude then from rate limiting
+
+### DNS controller manager
+
+The spec of a `DNSProvider` is extended by an optional `rateLimit` section.
```

Comment:
Will the `status` also be enhanced to track the daily limits?

Reply:
If the standard K8s client rate limiter is used (`NewTokenBucketRateLimiter`), you only know if a request would be allowed or not at the moment. There is no value representing "how much" of the daily limit is used. Anyway it is used in a way that there is no difference between requests/day or requests/second. I.e. the allowed requests are distributed equally over the day. And the rate limiter would be reset on restart, i.e. it is not persisted as this would much more laborious.
Instead I would rather see metrics for counting the requests per provider and for yes/no if the request limit is reached at at given point of time.